src/vma/sock/sockinfo_udp.cpp

```diff
@@ -1805,32 +1809,52 @@ bool sockinfo_udp::rx_input_cb(mem_buf_desc_t* p_desc, void* pv_fd_ready_array)
 		clock_gettime(CLOCK_REALTIME, &(p_desc->path.rx.sw_timestamp));
 	}
 
-	// In ZERO COPY case we let the user's application manage the ready queue
-	if (callback_retval != VMA_PACKET_HOLD) {
-		m_lock_rcv.lock();
-		// Save rx packet info in our ready list
-		m_rx_pkt_ready_list.push_back(p_desc);
-		m_n_rx_pkt_ready_list_count++;
-		m_rx_ready_byte_count += p_desc->path.rx.sz_payload;
-		m_p_socket_stats->n_rx_ready_pkt_count++;
-		m_p_socket_stats->n_rx_ready_byte_count += p_desc->path.rx.sz_payload;
-		m_p_socket_stats->counters.n_rx_ready_pkt_max = max((uint32_t)m_p_socket_stats->n_rx_ready_pkt_count, m_p_socket_stats->counters.n_rx_ready_pkt_max);
-		m_p_socket_stats->counters.n_rx_ready_byte_max = max((uint32_t)m_p_socket_stats->n_rx_ready_byte_count, m_p_socket_stats->counters.n_rx_ready_byte_max);
-		do_wakeup();
-		m_lock_rcv.unlock();
+	if (p_desc->path.rx.vma_polled) {
+		mem_buf_desc_t *tmp_p;
+		vma_completion_t* p_vma_completion;
+
+		p_vma_completion = (vma_completion_t*)pv_fd_ready_array;
+		p_vma_completion->packet.total_len = 0;
+
+		for(tmp_p = p_desc; tmp_p; tmp_p = tmp_p->p_next_desc) {
+			p_vma_completion->packet.buff_lst = (vma_buff_t*)tmp_p;
+			p_vma_completion->packet.buff_lst->next = (vma_buff_t*)tmp_p->p_next_desc;
+			p_vma_completion->packet.buff_lst->len = p_desc->path.rx.frag.iov_len;
+			p_vma_completion->packet.buff_lst->payload = p_desc->path.rx.frag.iov_base;
+			p_vma_completion->packet.total_len += tmp_p->path.rx.sz_payload;
+		}
+		p_vma_completion->events = VMA_POLL_PACKET;
+		p_vma_completion->src = p_desc->path.rx.src;
+		p_vma_completion->user_data = (uint64_t)m_fd_context;
+		p_vma_completion->packet.num_bufs = p_desc->n_frags;
```

Comment:
from looking at this I think `p_desc->path.rx.vma_polled = false` is not needed.
It will happen in `ring_sinple::rx_dispatch_packet()`

In case of MC adding this might even break... second socket will not handle it as vma_poll() logic.